### PR TITLE
[Inline Rename] Update the UI related to Copilot

### PR DESCRIPTION
**Pull request by veler**
Update the UI of Inline Rename in a manner that when Copilot is available, a click is necessary in order to start getting suggestions. Suggestions are then available in a drop down list.

smart-rename

image

image

image

image

---------------------

With updated icon:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
What's the key binding for this?  I don't think we can pass accessibility review if it always requires a mouse click.

**Comment by AmadeusW** [comment]
> What's the key binding for this? I don't think we can pass accessibility review if it always requires a mouse click.

Designers proposed Tab to set focus on the button, and Enter/Space to invoke it. 
Tab doesn't work, and we propose Ctrl+Space as a common way to invoke suggestion lists. 
Another suggestion is to use the Alt accelerator key, which is what powers the Include comments, Include strings checkboxes

**Comment by CyrusNajmabadi** [comment]
@AmadeusW can you show that on the hover of that button?  Thanks!

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 1]
```diff
@@ -197,7 +207,7 @@ private void Adornment_KeyDown(object sender, KeyEventArgs e)
                     // loop focus back to the first item that is focusable.
                     FrameworkElement lastItem = _viewModel.IsExpanded
```
Comment (AmadeusW) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:208]: @veler can you hook up Ctrl+Space to get suggestions?
`case Key.Space:`
`if (Keyboard.Modifiers == ModifierKeys.Ctrl)`

**Comment by vesarautiainen** [comment]
> > What's the key binding for this? I don't think we can pass accessibility review if it always requires a mouse click.
> 
> Designers proposed Tab to set focus on the button, and Enter/Space to invoke it. Tab doesn't work, and we propose Ctrl+Space as a common way to invoke suggestion lists. Another suggestion is to use the Alt accelerator key, which is what powers the Include comments, Include strings checkboxes

I don't mind adding Ctrl+Space to open the list. But I don't really understand why adding the generate button to the tab order is not doable. Isn't that just very standard way to navigate via keyboard (tab to focus to the next tab stop)? Just some examples to how I think it should work:

VS Code search box:
CleanShot 2023-12-19 at 11 41 45@2x

VS Code generate commit message:
CleanShot 2023-12-19 at 11 42 41@2x

VS Copilot chat: 
CleanShot 2023-12-19 at 11 43 52@2x

So is the problem that the generate button is inside the text box? Since if it was outside the text box I would expect that tab would move the focus to it. Wouldn't it be quite strange if you can see the button in the UI but there is no way to move the focus to it? I'm just quite adamant with this as I feel that to provide the best keyboard experience that button needs to be focusable.

**Comment by veler** [comment]
> What's the key binding for this? I don't think we can pass accessibility review if it always requires a mouse click.

@CyrusNajmabadi  We can use TAB followed by SPACE key to use it. The focus visual isn't there but I'm working on fixing it.
@AmadeusW, @vesarautiainen  considering this do we still need `Ctrl+Space` ?

File: src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1 +1,19 @@
-
+Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameUserInputPresenter
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt:1]: Could you change the types here to internal so they won't be public?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+          SelectionChanged="ComboBox_SelectionChanged"
+          Text="{Binding IdentifierText, UpdateSourceTrigger=PropertyChanged, Mode=TwoWay}"
+          ItemsSource="{Binding SmartRenameViewModel.SuggestedNames}"
+          StaysOpenOnEdit="True"
+          IsEditable="True"
+          Style="{StaticResource {x:Static vsfx:VsResourceKeys.ComboBoxStyleKey}}">
+    <ComboBox.Resources>
+        <ResourceDictionary>
+            <ResourceDictionary.MergedDictionaries>
+                <ResourceDictionary Source="../InlineRenameColors.xaml"/>
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml:21]: Is this needed?

**Comment by sharwell** [comment]
Can you include screenshots of the changes?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml [hunk 4]
```diff
@@ -0,0 +1,158 @@
+﻿<ComboBox x:Class="Microsoft.CodeAnalysis.InlineRename.UI.SmartRename.SmartRenameUserInputComboBox"
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml:1]: A general comment: I see there are some hard code colors in the control, could you verify it won't have any contrast problem in all the VS themes? (dark, light and blue)

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 5]
```diff
@@ -209,24 +219,19 @@ private void Adornment_KeyDown(object sender, KeyEventArgs e)
             }
         }
 
-        private void IdentifierTextBox_GotFocus(object sender, RoutedEventArgs e)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:212]: Why is this part removed?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyoutViewModel.cs [hunk 6]
```diff
@@ -60,13 +60,22 @@ internal class RenameFlyoutViewModel : INotifyPropertyChanged, IDisposable
             var smartRenameSession = smartRenameSessionFactory?.Value.CreateSmartRenameSession(_session.TriggerSpan);
             if (smartRenameSession is not null)
             {
-                SmartRenameViewModel = new SmartRenameViewModel(threadingContext, listenerProvider, smartRenameSession.Value);
+                SmartRenameViewModel = new SmartRenameViewModel(threadingContext, listenerProvider, smartRenameSession.Value, this);
                 SmartRenameViewModel.OnSelectedSuggestedNameChanged += OnSuggestedNameSelected;
+                SmartRenameViewModel.PropertyChanged += SmartRenameViewModel_PropertyChanged;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyoutViewModel.cs:65]: Unsubscribe this in `Dispose()`

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    private readonly Lazy<TextBox> _innerTextBox;
+    private Popup? _dropDownPopup;
+
+    internal SmartRenameUserInputComboBox(RenameFlyoutViewModel viewModel)
+    {
+        Requires.NotNull(viewModel, nameof(viewModel));
+        Requires.NotNull(viewModel.SmartRenameViewModel!, nameof(viewModel.SmartRenameViewModel));
+
+        InitializeComponent();
+        DataContext = viewModel;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:34]: Emm, I was originally expecting the Smart rename control would be bineded to smart rename view model.  (e.g smart rename control's data context would be smart rename view model, flyout's view model would be flyout view model, and you can interact the logic in VM)
Is there any special reason to use the general flyout view model here?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml (hunk 3 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml:21]: No, good catch! I struggled to apply the VS theme somehow.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs (hunk 7 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:34]: There's 1 binding done to `RenameFlyoutViewModel.IdentifierText`. But I just realized the `SmartRenameViewModel` has a `BaseViewModel` property to access it. I will switch binding to this and change data context to the smart rename view model. Thanks for pointing this out.

**Comment by veler** [comment]
> Can you include screenshots of the changes?

Here it is:
smart-rename

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs (hunk 5 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:212]: Oops.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml (hunk 4 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml:1]: Yes, I updated the PR description with some screenshots. Please take a look at them. Also, for your information, the theme of this control is copied then modified from the WPF theme we have internally in Visual Studio. The colors uses here are the same in our theme internally.

**Comment by vesarautiainen** [comment]
@veler : _We can use TAB followed by SPACE key to use it._ 

This is exactly what I had in mind. Not TAB + SPACE/ENTER at the same time. But just to tab to move the focus to the button and then SPACE/ENTER to invoke it.

**Comment by veler** [comment]
> @veler : _We can use TAB followed by SPACE key to use it._
> 
> This is exactly what I had in mind. Not TAB + SPACE/ENTER at the same time. But just to tab to move the focus to the button and then SPACE/ENTER to invoke it.

Great! We can still do this. Do we want to keep Ctrl+Space at the same time too?

**Comment by AmadeusW** [comment]
> Great! We can still do this. Do we want to keep Ctrl+Space at the same time too?

In the discussion outside of this PR, we concluded to keep Tab navigation as well as Ctrl+Space

I verified that the button correctly interacts with the Copilot related API.
I also verified that rename seems to work fine when Copilot is not installed with only one exception:

Pressing TAB causes focus to leave the TextBox (typing works fine, it's just that in current state of the change, the TextBox is grayed out instead of blue)
image

**Comment by AmadeusW** [comment]
Looks good from my point of view! Consider updating the main PR with up to date screenshots

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameControl.xaml.cs [hunk 8]
```diff
@@ -13,20 +11,9 @@ namespace Microsoft.CodeAnalysis.InlineRename.UI.SmartRename;
 /// </summary>
 internal sealed partial class SmartRenameControl : UserControl
 {
-    private readonly SmartRenameViewModel _smartRenameViewModel;
-
     public SmartRenameControl(SmartRenameViewModel viewModel)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameControl.xaml.cs:14]: Is this class still used somewhere? I thought since we introduced a combo box, this it no longer used

Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameControl.xaml.cs:14]: It is still used in the XAML

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 9]
```diff
@@ -51,20 +54,26 @@ internal partial class RenameFlyout : InlineRenameAdornment
...
+                RenameUserInput.GotFocus += RenameUserInput_GotFocus;
             };
 
             InitializeComponent();
 
+            RenameUserInputPresenter.Content = RenameUserInput;
+            RenameUserInput.PreviewKeyDown += RenameUserInput_PreviewKeyDown;
+
             // If smart rename is available, insert the control after the identifier text box.
             if (viewModel.SmartRenameViewModel is not null)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:78]: It seems now SmartRenameControl is only responsible for display status text? (Because all the suggestions are displayed by the combo box). Can we rename it to something like `SmartRenameStatusControl`?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 10]
```diff
@@ -86,6 +95,8 @@ internal partial class RenameFlyout : InlineRenameAdornment
             _ = DismissToolTipsAsync().CompletesAsyncOperation(token);
         }
 
+        internal IRenameUserInput RenameUserInput => _viewModel.SmartRenameViewModel is null ? _identifierTextBox.Value : _smartRenameUserInputComboBox.Value;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:98]: Is it possible to change this to `=> (IRenameUserInput)RenameUserInputPresenter.Content`?
My basic thought is to avoid the `_viewModel.SmartRenameViewModel is null` check if possible

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 11]
```diff
@@ -32,6 +33,8 @@ internal partial class RenameFlyout : InlineRenameAdornment
         private readonly IAsyncQuickInfoBroker _asyncQuickInfoBroker;
         private readonly IAsynchronousOperationListener _listener;
         private readonly IThreadingContext _threadingContext;
+        private readonly Lazy<RenameUserInputTextBox> _identifierTextBox;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:36]: Is it possible to change the code like
```
private readonly IRenameUserInput _renameUserInput;
```
And the field would be created based on if SmartRenameViewModel is null or not?
By have two fields here let me to think we have both input control at the same time

**Comment by Cosifne** [comment]
Is the `star` inco wider than a common button in VS? 
image

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs (hunk 11 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:36]: Actually, I could set the RenameUserInput property in the constructor, no need of fields anymore.

**Comment by veler** [comment]
> Is the `star` icon wider than a common button in VS? image

It was, but the icon and padding got changed since then. Is this better?
image

File: src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt [hunk 12]
```diff
@@ -1 +0,0 @@
-
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt:1]: I know it's boring 🙂but could you keep this file unchanged?

Comment (veler) [reply src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt:1]: Oh interesting, would have though Git find out there's no changes.

Comment (veler) [reply src/EditorFeatures/Core.Wpf/PublicAPI.Unshipped.txt:1]: Oh i see, there were 2 lines inside.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyoutViewModel.cs [hunk 13]
```diff
@@ -60,13 +60,22 @@ internal class RenameFlyoutViewModel : INotifyPropertyChanged, IDisposable
...
             }
 
             RegisterOleComponent();
         }
 
+        private void SmartRenameViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyoutViewModel.cs:71]: Could you check if `OnSuggestedNameSelected` and `SmartRenameViewModel_PropertyChanged` does a similar thing?
If so could they be merged? (Would prefer to use a special event instead of the general wpf property changing event to to this )

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 14]
```diff
@@ -60,20 +83,35 @@ internal sealed class SmartRenameViewModel : INotifyPropertyChanged, IDisposable
...
-        _smartRenameSession.GetSuggestionsAsync(_cancellationTokenSource.Token).CompletesAsyncOperation(listenerToken);
+        BaseViewModel = baseViewModel;
+        this.currentIdentifierText = baseViewModel.IdentifierText;
+
+        GetSuggestionsCommand = new DelegateCommand(OnGetSuggestionsCommandExecute, null, threadingContext.JoinableTaskFactory);
+    }
+
+    private void OnGetSuggestionsCommandExecute()
+    {
+        if (_getSuggestionsTask.Status is TaskStatus.RanToCompletion or TaskStatus.Faulted or TaskStatus.Canceled)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:109]: Is this method always invoked by the UI thread? if so could you add `_threadingContext.ThrowIfNotOnUIThread();` here?

Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:109]: it is on the UI thread indeed. I will update the PR :)

**Comment by Cosifne** [comment]
Also tag @ryzngard here. Could you give this a peek?
This is a request from the editor team, which adds another UI component (copilot suggestion) on top of the inline rename UI.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 15]
```diff
@@ -6,28 +6,33 @@
...
 namespace Microsoft.CodeAnalysis.InlineRename.UI.SmartRename;
 
 internal sealed class SmartRenameViewModel : INotifyPropertyChanged, IDisposable
 {
-    public static string GeneratingSuggestions => EditorFeaturesWpfResources.Generating_suggestions;
-
 #pragma warning disable CS0618 // Editor team use Obsolete attribute to mark potential changing API
     private readonly ISmartRenameSessionWrapper _smartRenameSession;
-#pragma warning restore CS0618 
+    public RenameFlyoutViewModel BaseViewModel { get; }
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:24]: Seem this doesn't need to be in the `#pragma warning disable` region?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 16]
```diff
@@ -41,6 +46,25 @@ internal sealed class SmartRenameViewModel : INotifyPropertyChanged, IDisposable
 
     public bool StatusMessageVisibility => _smartRenameSession.StatusMessageVisibility;
 
+    private string currentIdentifierText;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:49]: I believe it needs the `_` prefix

Comment (Cosifne) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:49]: So one more thought after the review of the xaml file.
It looks like you already did this `Text="{Binding BaseViewModel.IdentifierText, UpdateSourceTrigger=PropertyChanged, Mode=TwoWay}`

When user's selection changed, could it directly update the `Text` element value in xaml? And all the text change would be directly propaged to the `BaseViewModel.IdentifierText`
And we can get rid of both `CurrentIdentifierText` and `SelectedSuggestedName` here.

Comment (Cosifne) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:49]: Or you can set `BaseViewModel.IdentifierText` directly. (Depends on which one you prefer, I am fine with both)

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+        Assumes.NotNull(_dropDownPopup);
+        if ((e.Key == Key.Escape || e.Key == Key.Space || e.Key == Key.Enter) && _dropDownPopup.IsOpen)
+        {
+            _dropDownPopup.IsOpen = false;
+            SelectAllText();
+            e.Handled = true;
+        }
+    }
+
+    private void SuggestedNames_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:164]: Could you use `NotifyCollectionChangedEventArgs` and add the import?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 18]
```diff
@@ -221,56 +237,59 @@ private void Adornment_ConsumeMouseEvent(object sender, MouseButtonEventArgs e)
...
         }
 
         private void ToggleExpand(object sender, RoutedEventArgs e)
         {
             _viewModel.IsExpanded = !_viewModel.IsExpanded;
         }
 
+        private void RenameUserInput_GotFocus(object sender, RoutedEventArgs e)
+        {
+            this.RenameUserInput.SelectAllText();
```
Comment (ryzngard) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:254]: This feels like an implementation detail of `IRenameUserInput`, is there a reason the flyout handles this?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 19]
```diff
@@ -221,56 +237,59 @@ private void Adornment_ConsumeMouseEvent(object sender, MouseButtonEventArgs e)
...
+        }
+
         /// <summary>
         /// Respond to selection/cursor changes in the textbox the user is editing by
         /// applying the same selection to the textview that initiated the command
         /// </summary>
-        private void IdentifierTextBox_SelectionChanged(object sender, RoutedEventArgs e)
+        private void RenameUserInput_TextSelectionChanged(object sender, RoutedEventArgs e)
```
Comment (ryzngard) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:261]: Same here. This probably can be moved into the control itself since it already gets the `_viewModel` on creation

Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:261]: Mostly, keeping the control behavior in the control makes it easier to follow IMO

Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:261]: Also for `RenameUserInput_PreviewKeyDown` since it doesn't use anything unique to this control

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    private void ComboBox_GotKeyboardFocus(object sender, KeyboardFocusChangedEventArgs e)
+    {
+        e.Handled = true;
+    }
+
+    private void ComboBox_PreviewKeyUp(object sender, KeyEventArgs e)
+    {
+        if ((e.Key == Key.Up || e.Key == Key.Down) && Items.Count > 0)
```
Comment (ryzngard) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:116]: nit 
```suggestion
        if ((e.Key is Key.Up or Key.Down) && Items.Count > 0)
```

**Review by ryzngard — COMMENTED** [review]
Small comments. The conversation on the UI looks good to me. Make sure to do a pass with accessibility insights to catch what my human eyes can't, and request a manual run at some point to make sure everything is MAS compliant. I can't see any faults, but I don't know MAS by heart.

I give soft approval, since I don't want to unblock merging on behalf of the Roslyn team. Feel free to ping with any questions or follow up needed.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs (hunk 16 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:49]: You're right, I think this whole thing can be removed. Le me try.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs (hunk 18 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:254]: Mostly to centralize the text selection logic, so we don't have to spread "where do we select the text" in every implementation of `IRenameUserInput`.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs (hunk 19 above)
Comment (veler) [reply src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:261]: Same thing, I prefer keeping it here so we don't duplicate this logic into `SmartRenameUserInputComboBox` and `RenameUserInputTextBox`. I'm also not fan of putting it in `RenameUserInputPresenter` as I feel like we will spread logic across multiple area that might make the maintenance harder.

**Comment by veler** [comment]
> Small comments. The conversation on the UI looks good to me. Make sure to do a pass with accessibility insights to catch what my human eyes can't, and request a manual run at some point to make sure everything is MAS compliant. I can't see any faults, but I don't know MAS by heart.
> 
> I give soft approval, since I don't want to unblock merging on behalf of the Roslyn team. Feel free to ping with any questions or follow up needed.

Thanks! I just ran Accessibility Insight and it seems fine. Only thing not good is this button's icon in dark theme, which is already a known issue that existed before this PR:
image

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 21]
```diff
@@ -55,25 +59,41 @@ internal sealed class SmartRenameViewModel : INotifyPropertyChanged, IDisposable
...
         _threadingContext = threadingContext;
+        _listenerProvider = listenerProvider;
         _smartRenameSession = smartRenameSession;
         _smartRenameSession.PropertyChanged += SessionPropertyChanged;
-        var listener = listenerProvider.GetListener(FeatureAttribute.SmartRename);
 
-        var listenerToken = listener.BeginAsyncOperation(nameof(_smartRenameSession.GetSuggestionsAsync));
-        _smartRenameSession.GetSuggestionsAsync(_cancellationTokenSource.Token).CompletesAsyncOperation(listenerToken);
+        BaseViewModel = baseViewModel;
+        this.BaseViewModel.IdentifierText = baseViewModel.IdentifierText;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:83]: Is this line not needed?

**Review by Cosifne — APPROVED** [review]
One minor comment. 
🚀

**Review by arkalyanms — APPROVED** [review]
